### PR TITLE
Generic controller

### PR DESCRIPTION
**Pull request by SomtochiAma**

### CODE REVIEW

File: generic/main.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = addonsv1alpha1.AddToScheme(scheme) // <--------
```
Comment (justinsb) [inline generic/main.go:46]: Nit: We can remove the `<---` marker :-)

File: generic/main.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		setupLog.Error(err, "unable to destructure unstructured")
+		os.Exit(1)
+	}
+
+	if len(genericList.Items) == 0 {
+		fmt.Fprint(os.Stderr, "Please create a `Generic` resource\n")
+		os.Exit(1)
+	}
+
+	for i := range genericList.Items {
```
Comment (justinsb) [inline generic/main.go:102]: Nit: I think we should have a TODO to say "add a watch on Generic, respond to changes" or something like that

File: generic/controllers/generic_controller.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	declarative.Reconciler
+	GVK     schema.GroupVersionKind
+	Channel string
+}
+
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=generics,verbs=get;list;watch;create;update;patch;delete
```
Comment (justinsb) [inline generic/controllers/generic_controller.go:39]: I actually don't think we need these rbac permissions.  I guess we want get/list/watch on generics, but you could also move than to main.go.

File: generic/controllers/generic_controller.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	); err != nil {
+		return err
+	}
+
+	c, err := controller.New(strings.ToLower(r.GVK.Kind)+"-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Generic
```
Comment (justinsb) [inline generic/controllers/generic_controller.go:79]: Nit: I think this should say "Watch for changes to the objectKind" or something like that

**Comment by justinsb** [comment]
This is looking really good!  We do need the tools to generate the RBAC & the manifests, but then this is going to be a really nice experience for users!

File: generic/controllers/generic_controller.go (hunk 3 above)
Comment (SomtochiAma) [reply generic/controllers/generic_controller.go:40]: Is there any difference, moving it to `main.go`

**Comment by cmoulliard** [comment]
Can you please add a `README.md` file under the tools in order to explain how to use them with some concrete examples ?

File: generic/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot
+WORKDIR /
+COPY --from=builder /workspace/manager .
+COPY addons/ addons/
```
Comment (cmoulliard) [inline generic/Dockerfile:25]: Error. The line `COPY addons/ addons/` should be changed to ` COPY channels/ channels/` otherwise docker build will fail.

**Comment by cmoulliard** [comment]
I got this error when I run the generic controller under a kind cluster using the command `make deploy`

```
kc logs generic-controller-manager-5d9cf5db7b-mnfc2 -n generic-system -c manager
2020-08-06T10:27:56.581Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
Please create a `Generic` resource
```
Why should we do then deploy first a Generic CRD before to start the controler ?

**Comment by SomtochiAma** [comment]
> I got this error when I run the generic controller under a kind cluster using the command `make deploy`
> 
> ```
> kc logs generic-controller-manager-5d9cf5db7b-mnfc2 -n generic-system -c manager
> 2020-08-06T10:27:56.581Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
> Please create a `Generic` resource
> ```
> 
> Why should we do then deploy first a Generic CRD before to start the controler ?

The generic controller is different because it is used as a general controller for other simple addons that don't require any extra configuration. 
The Generic resource is used to tell the generic controller these addons are and where to find their channels

**Comment by SomtochiAma** [comment]
> Can you please add a `README.md` file under the tools in order to explain how to use them with some concrete examples ?

I will definitely do so

**Comment by cmoulliard** [comment]
> The Generic resource is used to tell the generic controller these addons are and where to find their channels

Maybe then review the message logged to make it more explicit about root cause of the error and what is it needed to be done to avoid it. As the controller will crash, then we must find a way to make it healthy if a Generic CR has been deployed ...

**Comment by SomtochiAma** [comment]
> Maybe then review the message logged to make it more explicit about root cause of the error and what is it needed to be done to avoid it. 

I changed the message and added a REAME.md for the tools. Please review whenever you can! Thanks

**Comment by justinsb** [comment]
This is really cool and it sounds like people want to start trying it out, so let's merge it and iterate with further improvements!

/approve
/lgtm

**Pull request merged**